### ISSUE TITLE
Figure out what it will take to build the BCL with the interpreter on CI

### ISSUE THREAD

**Issue by luhenry** [mono/mono#7418]
We want to build the BCL and the tests - on top of running the latter - with the interpreter on its CI lanes. We must try to use Roslyn, but should fallback to MCS if there are too many failures.

This would allow us:
 - Augmented trust of our CI: if we are able to run Roslyn or MCS to compile the BCL, this is a good sign that the interpreter is solid; it already allows us to uncovers JIT bugs
 - Easier testing for mixed-mode: by using the already AOTted core libraries and executables (mscorlib, System, roslyn/mcs, etc.) to compile the BCL with the interpreter, we would test mixed-mode.

First, we need to assess the situation and figure out what's broken and how much will be needed to fix it.

**Comment by luhenry** [comment]
Compiling the BCL with MCS doesn't crash all the time, so it should be possible to use the interpreter to compile the BCL and tests, and run the tests.

**Comment by marek-safar** [comment]
This adds almost no value. We are already running mcs tests suite which covers many more cases than running mcs for BCL.

We need to run csc compiled BCL tests with interpreter. That's significantly more useful target.

**Comment by luhenry** [comment]
@marek-safar csc is a much bigger target to support than mcs, possibly putting it many month away from doable. mcs is the only solution over csc that would allow us to guarantee we do not regress the interpreter when compiling the BCL. Discarding it as "does not add value" is just pushing aside the existing context.

**Comment by marek-safar** [comment]
I didn't say we have to run csc using the interpreter. I just don't see little value in using the interpreter to build BCL. What does it add us on top of what we have today? If you can run mcs test suite using the interpreter we can almost certainly build BCL using mcs. There are not special IL patterns for the different BCL libraries used in mcs

**Comment by marek-safar** [comment]
I disagree, the mcs code is pretty self-contained and uses almost no BCL APIs it might use different IL patterns but they should all be hit by mcs tests. We might want to run mscorlib compilation using the interpreter but apart from that I'd focus elsewhere, there are many more tests we could run which would give us significantly better coverage in less time.

**Comment by luhenry** [comment]
@marek-safar the added value is the same as compiling the BCL with a just-compiled BCL (the `build` profile), or compiling the BCL with the just-compiled runtime: mcs and csc are complicated programs that catches bugs which are much more complex that what most unit tests can do, so compiling the BCL with the Interpreter (or the JIT, or the `build` BCL for the same matter) has an **immense** added value for the Interpreter. Arguing that it adds no value is the same as arguing that the `build` profile does not add value (it's not useful for boostraping, `basic` already fills that role), and compiling the BCL with the just-built runtime does not add value either in the presence of a system mono (ie, it's a bad argument)

We are already looking at enabling more tests (see the System, System.Core, corlib test suites), and building the BCL with the interpreter is just one of these steps to the end goal. Also, compiling the BCL with the interpreter is **already** working as we saw in https://github.com/mono/mono/pull/7722, the only thing that is blocking are 5 failing tests (failures are specifically due to mcs and not the interpreter).

**Comment by luhenry** [comment]
The discussion can be summarized in two possible plan of actions:
 1. fix mcs and build the BCL with the interpreter+mcs, then later on fix the interpreter to support roslyn and build the BCL with the interpreter+roslyn.
 2. fix the interpreter for roslyn and build the BCL with the interpreter+roslyn

Both plans have the same end goal - building the BCL with interpreter+roslyn -, the first one allows us to reduce the chances of introducing regressions sooner rather than later but adds the requirement of fixing mcs, while the second one just postpone building the BCL with the interpreter.

So the question is: @marek-safar how much work would it be to fix mcs to build the BCL and have it pass the BCL test suite? (note that this isn't related to the interpreter or the JIT, but only to mcs, since the BCL test suite fails with both JIT and interpreter if compiled with mcs)

**Comment by marek-safar** [comment]
We have that already mcs can build BCL we test that on every PR so nothing to fix there. I still see little value of running that using interpreter compiling hundreds of Facades assemblies just to prove that same mcs code path work over and over again but if we can run it in reasonable time on every PR.

The scenario where mcs is used to compile BCL tests is not something we want to spend time on.
